### PR TITLE
Update scoring function interface

### PR DESCRIPTION
**Pull request by adamlm**
# PR Details
## Description

The old function for scoring tracks and objects has been replaced with one that is a little more user friendly. Now, the scoring map only contains matches that have a valid score. Using `std::optional` has been restricted to only the metric function object.

The unit tests have been updated accordingly.

The Visitor class has been moved out of utils to its own file. Having a utils file is in general bad practice. Instead, the functionality should be put into specific files, even if they are small.

There is now a common visitors file to house visitors common to several components. Currently only the uuid visitor is in it.

## Related GitHub Issue

Closes usdot-fhwa-stol/cooperative-perception-core#36

## Motivation and Context

As the library gets closer to being feature complete, having a predictable and unified interface will help developers interact with it.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by MishkaMN — CHANGES_REQUESTED** [review]

File: include/cooperative_perception/scoring.hpp [hunk 1]
```diff
@@ -57,20 +56,26 @@ constexpr utils::Visitor mahalanobis_distance_visitor{ [](const auto& track,
   }
 } };
 
-template <typename DistanceVisitor>
-auto score_tracks_and_objects(const std::vector<TrackType>& tracks, const std::vector<DetectionType>& objects,
-                              const DistanceVisitor& distance_visitor)
+using ScoreMap = std::map<std::pair<std::string, std::string>, float>;
```
[the remark this answers is not in the document]
Comment (adamlm) [reply include/cooperative_perception/scoring.hpp:59]: We had some issue with the hashing for `std::unordered_map`, so we switched to this. I thought `std::pair<std::string, std::string>` already had a supported hashing operation in the standard library, but maybe not.

Either way, I think we can stick with this and change it later if necessary. If the standard library cannot hash that key type, we can always use `boost::hash`.

File: include/cooperative_perception/visitor.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  using Base::operator()...;
+};
+
+/**
+ * @brief Type deduction hint for the Visitor constructor
+ */
+template <typename... T>
+Visitor(T&&... t) -> Visitor<T...>;
+
+}  // namespace cooperative_perception
```
[the remark this answers is not in the document]
Comment (adamlm) [reply include/cooperative_perception/visitor.hpp:47]: Some code outside this PR's scope still uses it. I think putting this change into a separate issue would be better for documentation. What do you think?

File: include/cooperative_perception/scoring.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (MishkaMN) [reply include/cooperative_perception/scoring.hpp:59]: I see okay. I am good with it then

File: include/cooperative_perception/visitor.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamlm) [reply include/cooperative_perception/visitor.hpp:47]: After internal discussion with @MishkaMN, we will remove `utils::Visitor` in this PR.

**Review by MishkaMN — APPROVED** [review]

**Pull request merged**